REGRESSION ( Sonoma?): [ Sonoma release ] TestWebKitAPI.WKInspectorExtension.EvaluateScriptOnPage is a consistent failure

rcaliman-apple opened the pull request:
#### 3583b2ac96f10694075cac1324cf5b14a41d77cb
<pre>
REGRESSION ( Sonoma?): [ Sonoma release ] TestWebKitAPI.WKInspectorExtension.EvaluateScriptOnPage is a consistent failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=263447">https://bugs.webkit.org/show_bug.cgi?id=263447</a>
<a href="https://rdar.apple.com/117265270">rdar://117265270</a>

Reviewed by Timothy Hatcher.

The API test loads a page from an URL, opens Web Inspector for it, registers a Web Inspector Web Extension
and immediately sends a request to evaluate some JavaScript in the main frame and subframes for that URL.

The test fails because of a race condition: the objects that represent the frame and execution context
within Web Inspector frontend are not necessarily available when the evaluation request comes in.

The missing objects result in early returns with error codes:
`WI.WebInspectorExtension.ErrorCode.InvalidRequest` or `WI.WebInspectorExtension.ErrorCode.ContextDestroyed`.

Fixing the race means waiting for the corresponding `WI.Frame` and `WI.ExecutionContext` instances
to be available before attempting to use them. A simple retry approach with a timeout works fine.

---

A `WI.Frame` is created in response to initializing an `InspectorPageAgent`
and receiving a response to the `getResourceTree()` command.

A `WI.ExecutionContext` is created asynchronously in response to the &quot;executionContextCreated&quot; event
dispatched by `PageRuntimeAgent`.

The test failed consistently on Release builds but not on Debug builds. The slower Debug builds likely gave
enough time for the Web Inspector frontend to be created and populated with the necessary objects.

---

Why is Web Inspector frontend involved in an Web Extensions API test? Good question!

The call to evaluate a script from a Web Inspector web extension in the page context goes down this path:

`WebInspectorUIExtensionController::evaluateScriptForExtension` (C++)
    -&gt; `InspectorFrontendAPIDispatcher::dispatchCommandWithResultAsync()` (C++, Web Inspector backend)
        -&gt; `WI.WebInspectorExtensionController.evaluateScriptForExtension()` (JS, Web Inspector frontend)
            -&gt; `InspectorRuntimeAgent::evaluate()` (C++, Web Inspector backend)

The result of the evaluation gets ferried on the same route all the way back.

* LayoutTests/inspector/unit-tests/retryuntil.html: Added.
* LayoutTests/inspector/unit-tests/retryuntil-expected.txt: Added.
* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
* Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:
(WI.WebInspectorExtensionController.prototype.async evaluateScriptForExtension):
(WI.WebInspectorExtensionController.prototype.evaluateScriptForExtension): Deleted.

Canonical link: <a href="https://commits.webkit.org/271925@main">https://commits.webkit.org/271925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c97c89698f332a6e78881a2a543a5fb30639fd5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8650 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31302 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32493 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27111 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10840 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5906 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27156 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30280 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7255 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25578 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6190 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6360 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26667 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/autoplay-allowed-by-feature-policy-attribute.https.sub.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33829 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27350 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27099 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32528 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6294 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4465 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30324 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8034 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/26465 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7125 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7039 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6821 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:
```diff
@@ -125,15 +125,32 @@ WI.WebInspectorExtensionController = class WebInspectorExtensionController exten
+            }
+
+            return null;
+        }
```
This seems like a useful thing to put in Base as a standalone helper, like `Debouncer` and `Throttler`.

rcaliman-apple replied:
It's a tiny utility, so perhaps it can go in [Source/WebInspectorUI/UserInterface/Base/Utilities.js](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Base/Utilities.js) instead of having its own dedicated file.

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Utilities.js:
```diff
@@ -1866,6 +1866,24 @@ function insertObjectIntoSortedArray(object, array, comparator)
     array.splice(insertionIndexForObjectInListSortedByFunction(object, array, comparator), 0, object);
 }
 
+async function retryUntilTruthy(fn, {delay, retries} = {})
```
this should have tests

NIT: id rename this just `retryUntil` as the "Truthy" part is implicit/redundant.  also maybe a name like `callUntil` is more apt since it's possible it may only be called once and therefore not ever "retried"

Style: please use something more descriptive than `fn`

rcaliman-apple replied:
> this should have tests

Added. 

> NIT: id rename this just `retryUntil` as the "Truthy" part is implicit/redundant. also maybe a name like `callUntil` is more apt since it's possible it may only be called once and therefore not ever "retried"

Went with `retryUntil`. I tried `callUntil` but I felt it didn't communicate the intent as clearly a "retry" when read in context.

> Style: please use something more descriptive than `fn`

Used "predicate"

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Utilities.js:
```diff
@@ -1866,6 +1866,24 @@ function insertObjectIntoSortedArray(object, array, comparator)
+    delay ??= 100;
+    let count = 0;
+
+    while (count < retries) {
```
id have this be a `for (let i = 0; i < retries; ++i) {` instead since that's basically what it's doing

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Utilities.js:
```diff
@@ -1866,6 +1866,24 @@ function insertObjectIntoSortedArray(object, array, comparator)
+        await Promise.delay(delay);
+    }
+
+    return null;
```
should we have some sort of `console.assert(false, "exceeded retries", fn, retries, delay);`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:
```diff
@@ -125,15 +125,15 @@ WI.WebInspectorExtensionController = class WebInspectorExtensionController exten
         }
 
-        let frame = this._frameForFrameURL(frameURL);
+        let frame = await retryUntilTruthy(() => { return this._frameForFrameURL(frameURL); });
```
Style: you can omit the `{ return` and `}`
```
let frame = await callUntil(() => this._frameForFrameURL(frameURL));
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:
```diff
@@ -149,7 +149,7 @@ WI.WebInspectorExtensionController = class WebInspectorExtensionController exten
         }
 
-        let evaluationContext = frame.pageExecutionContext;
+        let evaluationContext = await retryUntilTruthy(() => { return frame.pageExecutionContext; });
```
ditto
```
let evaluationContext = await callUntil(() => frame.pageExecutionContext);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Utilities.js:
```diff
@@ -1866,6 +1866,24 @@ function insertObjectIntoSortedArray(object, array, comparator)
 
+async function retryUntilTruthy(fn, {delay, retries} = {})
+{
+    retries ??= 30;
```
do we ever actually need 30 retries?  or was this just a number you picked randomly?

i am slightly concerned that this may become flaky if things get slower in the future 🤔

perhaps we should have it actually be `Infinity`?  would that actually be able to cause issues?

rcaliman-apple replied:
> do we ever actually need 30 retries? or was this just a number you picked randomly?

Yes, it was picked at random. I increased it to 100 retries * 100ms delay for a maximum of 10 seconds of retry, which should be reasonable for the default case. 🤞 

> i am slightly concerned that this may become flaky if things get slower in the future 🤔
> 
> perhaps we should have it actually be `Infinity`? would that actually be able to cause issues?

We do want this to eventually timeout and return `null` so there's meaningful a trace to debug.

Tests have their own timeout mechanism, but this method could be used in another context without an escape hatch where an infinite wait may be harder to debug.

Where `Infinity` is truly desired, it can be configured: `retryUntil(predicate, {retries: Infinity})`

rcaliman-apple committed:
REGRESSION ( Sonoma?): [ Sonoma release ] TestWebKitAPI.WKInspectorExtension.EvaluateScriptOnPage is a consistent failure
https://bugs.webkit.org/show_bug.cgi?id=263447
rdar://117265270

Reviewed by Timothy Hatcher.

The API test loads a page from an URL, opens Web Inspector for it, registers a Web Inspector Web Extension
and immediately sends a request to evaluate some JavaScript in the main frame and subframes for that URL.

The test fails because of a race condition: the objects that represent the frame and execution context
within Web Inspector frontend are not necessarily available when the evaluation request comes in.

The missing objects result in early returns with error codes:
`WI.WebInspectorExtension.ErrorCode.InvalidRequest` or `WI.WebInspectorExtension.ErrorCode.ContextDestroyed`.

Fixing the race means waiting for the corresponding `WI.Frame` and `WI.ExecutionContext` instances
to be available before attempting to use them. A simple retry approach with a timeout works fine.

---

A `WI.Frame` is created in response to initializing an `InspectorPageAgent`
and receiving a response to the `getResourceTree()` command.

A `WI.ExecutionContext` is created asynchronously in response to the "executionContextCreated" event
dispatched by `PageRuntimeAgent`.

The test failed consistently on Release builds but not on Debug builds. The slower Debug builds likely gave
enough time for the Web Inspector frontend to be created and populated with the necessary objects.

---

Why is Web Inspector frontend involved in an Web Extensions API test? Good question!

The call to evaluate a script from a Web Inspector web extension in the page context goes down this path:

`WebInspectorUIExtensionController::evaluateScriptForExtension` (C++)
    -> `InspectorFrontendAPIDispatcher::dispatchCommandWithResultAsync()` (C++, Web Inspector backend)
        -> `WI.WebInspectorExtensionController.evaluateScriptForExtension()` (JS, Web Inspector frontend)
            -> `InspectorRuntimeAgent::evaluate()` (C++, Web Inspector backend)

The result of the evaluation gets ferried on the same route all the way back.

* LayoutTests/inspector/unit-tests/retryuntil.html: Added.
* LayoutTests/inspector/unit-tests/retryuntil-expected.txt: Added.
* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
* Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:
(WI.WebInspectorExtensionController.prototype.async evaluateScriptForExtension):
(WI.WebInspectorExtensionController.prototype.evaluateScriptForExtension): Deleted.

Canonical link: https://commits.webkit.org/271925@main
